Camel-quarkus-datasonnet extension - resolves Issue #2078

javaduke opened the pull request:
<!-- Uncomment and fill this section if your PR is not trivial
[ ] An issue should be filed for the change unless this is a trivial change (fixing a typo or similar). One issue should ideally be fixed by not more than one commit and the other way round, each commit should fix just one issue, without pulling in other changes.
[ ] Each commit in the pull request should have a meaningful and properly spelled subject line and body. Copying the title of the associated issue is typically enough. Please include the issue number in the commit message prefixed by #.
[ ] The pull request description should explain what the pull request does, how, and why. If the info is available in the associated issue or some other external document, a link is enough.
[ ] Phrases like Fix #<issueNumber> or Fixes #<issueNumber> will auto-close the named issue upon merging the pull request. Using them is typically a good idea.
[ ] Please run mvn process-resources -Pformat (and amend the changes if necessary) before sending the pull request.
[ ] Contributor guide is your good friend: https://camel.apache.org/camel-quarkus/latest/contributor-guide.html
-->

aldettinger commented:
@javaduke Great to see that an extension is almost there :)

I wonder if the build issue on CI would be linked to inverting the order of bom imports. That switch might have consequences, we'll need to find an answer here.

Apart from that, a curiosity, the PR embeds dslibs.jar in the system under test for "datasonnet functionaly testing" purpose. Is it really needed to embed this ? For instance, couldn't it just be a test scoped dependency ?

javaduke commented:
I think I found a better way to avoid pulling the old scala library. I'm testing it now.
The situation with `dslibs.jar` is different. This is just a resource inside the jar and the test verifies if datasonnet is able to read a library stored in a jar. With JVM tests I can just add the jar as a configuration parameter for the surefire plugin. This doesn't work for native tests though. I tried adding it to the pom as a test scope, but the problem is that this jar is not really a maven artifact, it's not in any maven repo, and `systemPath` cannot be used with any scope other than `system`. So I'm not sure if there's any other option to include this in the classpath for native tests.

aldettinger commented:
Hey, @javaduke it's really lovely to see contributions :)

So, concerning the bom import order. It's better to change only at datasonnet level. I'm under the impression that it could still generate some issues with respect to enforcers and quarkus-platform. We'll see.

Now, regarding the dslibs.jar, I see a bit more. At this stage, I would merely remove this part from testing. We try to avoid jar in sources as far as possible. Plus, I would assume most camel-quarkus users would define there libs in their own app ? And also, some tricks might be needed in native mode. Anyway, that could be another ticket.

With ther2.8 release underway, it's maybe a better idea to review this PR next week with the embedded jar removed. That would already be a good first step :) Keep up the good work :+1:

javaduke commented:
Ah, I was hoping to squeeze the data sonnet into the 2.8.0 release, because our work really depends on it.
Anyway, I removed the named imports test (no big deal, it is tested in `camel-datasonnet` anyway).

aldettinger commented on extensions/datasonnet/deployment/src/main/java/org/apache/camel/quarkus/component/datasonnet/deployment/DatasonnetProcessor.java:
```diff
@@ -0,0 +1,59 @@
+    void addDependencies(BuildProducer<IndexDependencyBuildItem> indexDependency) {
+        indexDependency.produce(new IndexDependencyBuildItem("com.datasonnet", "datasonnet-mapper"));
+        indexDependency.produce(new IndexDependencyBuildItem("org.scala-lang", "scala-library"));
+        indexDependency.produce(new IndexDependencyBuildItem("org.scala-lang.modules", "scala-collection-compat_2.13"));
```
On scala version upgrade tests should capture this, so we are fine here IMO.

aldettinger replied:
Sorry, I was thinking aloud what would happen in case of scala version upgrade :)
And I think, the tests would fail, then we detect that some dependencies are not indexed, and then we update those lines of code. If you agree with my understanding, then I think we should keep this code as is.

aldettinger replied:
Reading again, I think my understanding is correct here. So we can keep like this.

aldettinger commented on extensions/datasonnet/pom.xml:
```diff
@@ -0,0 +1,58 @@
+                <type>pom</type>
+                <scope>import</scope>
+            </dependency>
+        </dependencies>
```
@ppalaga @jamesnetherton This PR changes the quarkus and cq pom import orders. I think there will be issues, especially with the plaform. Do you agree that we need to find another way or it could be fine ?

ppalaga replied:
Yes, `quarkus-bom` should always be before `camel-quarkus-bom`. Please remove the whole dependencyManagement section here, because the imports are there in the parent `camel-quarkus-extensions`.

aldettinger commented on extensions/datasonnet/runtime/pom.xml:
```diff
@@ -0,0 +1,119 @@
+
+    <properties>
+        <camel.quarkus.jvmSince>2.8.0</camel.quarkus.jvmSince>
+        <camel.quarkus.nativeSince>2.8.0</camel.quarkus.nativeSince>
```
We may need to update the versions at some point in time.

javaduke replied:
Done.

javaduke replied:
Done.

aldettinger replied:
Great. Many thanks.

aldettinger commented on extensions/datasonnet/runtime/pom.xml:
```diff
@@ -0,0 +1,119 @@
+    <parent>
+        <groupId>org.apache.camel.quarkus</groupId>
+        <artifactId>camel-quarkus-datasonnet-parent</artifactId>
+        <version>2.8.0-SNAPSHOT</version>
```
Also here, the main branch will move to 2.9.0-SNAPSHOT in a moment.

javaduke replied:
Done.

javaduke commented:
Sorry, I'm confused, would you like me to make any changes here?

aldettinger commented:
Hey @javaduke, there is another PR that will update the version to 2.9.0-SNAPHOT https://github.com/apache/camel-quarkus/pull/3696. So once, #3696 is merged, then we would need to update the versions here.

Generally speaking the PR is good. The only issue I could see is the bom re-odering, I've asked help/confirmation about this. Still people are very busy. Worst case, I should be able to investigate further later on.

javaduke commented:
OK, np, I'll keep an eye on the PR

javaduke commented:
I bumped versions to 2.9.0 and committed again.

ppalaga commented on extensions/datasonnet/deployment/src/main/java/org/apache/camel/quarkus/component/datasonnet/deployment/DatasonnetProcessor.java:
```diff
@@ -0,0 +1,59 @@
+
+    @BuildStep
+    void process(CombinedIndexBuildItem combinedIndexBuildItem,
+            BuildProducer<ReflectiveClassBuildItem> reflectiveClasses,
```
`combinedIndexBuildItem`and `reflectiveClasses` seem to be redundant. If not, please add a comment explaining what's their purpose.

javaduke replied:
Done.

javaduke commented:
All done.

ppalaga commented:
Thanks @javaduke, let's wait for the CI.

javaduke commented:
Hmm, I'm not sure why is it failing, complains about generated flattened poms...

javaduke commented:
Any updates on this? Is there anything else that is needed to be fixed?

aldettinger commented on integration-tests/datasonnet/pom.xml:
```diff
@@ -0,0 +1,176 @@
+        </plugins>
+    </build>
+
+    <dependencyManagement>
```
Why is this dependencyManagement section needed ? It may lead to issue.

javaduke commented:
It fails with complains that there are uncommitted changes:
```
Run [[ -z $(git status --porcelain | grep -v antora.yml) ]] || { echo 'There are uncommitted changes'; git status; exit 1; }
[10](https://github.com/apache/camel-quarkus/runs/6080356797?check_suite_focus=true#step:10:10)
There are uncommitted changes
[11](https://github.com/apache/camel-quarkus/runs/6080356797?check_suite_focus=true#step:10:11)
HEAD detached at pull/3686/merge
[12](https://github.com/apache/camel-quarkus/runs/6080356797?check_suite_focus=true#step:10:12)
Changes not staged for commit:
[13](https://github.com/apache/camel-quarkus/runs/6080356797?check_suite_focus=true#step:10:13)
  (use "git add <file>..." to update what will be committed)
[14](https://github.com/apache/camel-quarkus/runs/6080356797?check_suite_focus=true#step:10:14)
  (use "git restore <file>..." to discard changes in working directory)
[15](https://github.com/apache/camel-quarkus/runs/6080356797?check_suite_focus=true#step:10:15)
	modified:   poms/bom/src/main/generated/flattened-reduced-pom.xml
[16](https://github.com/apache/camel-quarkus/runs/6080356797?check_suite_focus=true#step:10:16)
	modified:   poms/bom/src/main/generated/flattened-reduced-verbose-pom.xml
[17](https://github.com/apache/camel-quarkus/runs/6080356797?check_suite_focus=true#step:10:17)

[18](https://github.com/apache/camel-quarkus/runs/6080356797?check_suite_focus=true#step:10:18)
no changes added to commit (use "git add" and/or "git commit -a")
```

I have the same issue with the CXF extension CI build, and I can't reproduce it locally, so I have no idea what is wrong.

javaduke commented:
I wonder if generated poms should not be in the repository, since they are generated by maven, maybe deleting them fixes the problem?

ppalaga commented on pom.xml:
```diff
@@ -127,6 +128,7 @@
         <retrofit.version>2.5.0</retrofit.version>
         <saaj.impl.version>1.5.3</saaj.impl.version>
         <scala-2.12.version>2.12.15</scala-2.12.version><!-- Spark -->
+        <scala.version>2.13.3</scala.version><!-- Datasonnet -->
```
Scala version 2.12 is currently dictated by Quarkus. We could override it in our BOM, but any application generated by code.quarkus.io would get the version from Quarkus BOM anyway. 
Hence we have only two options here: (a) ask Quarkus to upgrade to Scala 2.13 or (b) make datasonnet work with 2.12.
I see this as the main issue of this PR.

javaduke replied:
Sorry, I don't think it's possible to make Datasonnet work with 2.12...so I guess we're in the holding pattern until Quarkus upgrades to 2.13

ppalaga replied:
OK, thanks for confirming that it cannot work with Scala 2.12. There is this quite old quest to upgrade in Quarkus: https://github.com/quarkusio/quarkus/issues/10025 I have added a comment there.

ppalaga replied:
Scala 2.13 in Quarkus is underway https://github.com/quarkusio/quarkus/pull/25106

ppalaga reviewed (changes requested):
Please see inline. We need to figure out how we proceed with the Scala version.

ppalaga commented:
> I wonder if generated poms should not be in the repository, since they are generated by maven, maybe deleting them fixes the problem?

We track the generated poms in git to have some long term auditing. E.g. in this PR, the changes in those file clearly show that you attempt to override scala libs maintained by Quarkus.

ppalaga commented on integration-tests/datasonnet/pom.xml:
```diff
@@ -0,0 +1,176 @@
+            <artifactId>rest-assured</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.skyscreamer</groupId>
+            <artifactId>jsonassert</artifactId>
+            <version>1.5.0</version>
+            <scope>test</scope>
+        </dependency>
```
Comparing the JSON documents as strings would not work here?

javaduke replied:
No, because order is not guaranteed, and formatting may be off.

javaduke replied:
Yup, this works, thank you!

aldettinger commented on integration-tests/datasonnet/pom.xml:
```diff
@@ -0,0 +1,176 @@
+        <dependency>
+            <groupId>org.apache.camel.quarkus</groupId>
+            <artifactId>camel-quarkus-datasonnet</artifactId>
+            <exclusions>
+                <exclusion>
+                    <groupId>org.scala-lang</groupId>
+                    <artifactId>scala-library</artifactId>
+                </exclusion>
+            </exclusions>
```
Indeed, this exclusion might not be needed. Logically, it could be possible to check by running `mvn depedency:tree` with and without the exclusion. The same dependency:tree should be resolved in both cases.

ppalaga commented:
@javaduke if you could please `rebase` and `push -f` rather than merge main. It is much easier for keeping an overview what is the actual state of your topic branch. Also in this project we prefer single commit per logical change. Future readers of the history might not be interested in all intermediate changes. Thanks for your great work!

ppalaga commented:
The Scala upgrade in https://github.com/quarkusio/quarkus/pull/25106 is ready to be merged. You might want to build locally against that PR, check that it works for you and give feedback on that PR if necessary.

javaduke commented:
I built quarkus locally and changed the version in the main pom.xml to 999-SNAPSHOT, but for some reason the flattened poms still pull 2.8.0.Final, and the resulting scala version is 2.12...

aldettinger commented:
@javaduke I don't know exactly what is the issue at hand. Still, the quarkus-main branch may come in handy as it points to [quarkus last SNAPSHOT](https://github.com/apache/camel-quarkus/blob/quarkus-main/pom.xml#L59). Maybe relocating to this branch would be a good path ?

javaduke commented:
Yes, I just tried it, same result - in my dependency tree I see `io.quarkus:quarkus-core:jar:999-SNAPSHOT:compile` which is what I expected, but I also see ` org.scala-lang:scala-library:jar:2.12.13:compile`.

aldettinger commented:
@javaduke We are progressing, that's great :+1:  What about squashing all your commits into a single and opening another PR against the quarkus-main branch ? With such a PR, I feel we would be 2 or 3 details far from merging.

javaduke commented:
Closing this PR since the new one was opened per @aldettinger request: https://github.com/apache/camel-quarkus/pull/3757/
